### DISCUSSION TITLE
Incompatible TOTP-storage formats?

### DISCUSSION THREAD

**Discussion by borque** [keepassxreboot/keepassxc#8094] [Q&A] [answered]
Hello!
I plan to store my growing number of TOTP-secrets in a seperate KeePass-database.
While planning this I did some research and noticed there might be an incompatibility in the way TOTP-information is stored?!

KeePassXC seems to store them in a single field called OTP where all information is coded in URL-style.
Many other Implementations (such as Strongbox, Keepassium, the KeeTrayTOTP plugin for KeePass) seem to store them in to different fields TOTPseed and TOTPsettings.

Indeed I was unable to get TOTP-codes from a test-database created with KeePassXC 2.7 in KeePass 2.5 with KeeTrayTOTP-plugin. And the  KeeTrayTOTP-plugin refused to set the fields for seed and settings to the same databasefield.

Am I making something wrong? Is there a way to store TOTP data in a compatible and interoperable way?

Thanks for help!

**Comment by droidmonkey** [comment, marked as answer]
There is no standard way to store until recently when KeePass project implemented a multi-attribute method. The otp field is supported by all major mobility apps, but not KeePass. You will have to configure your TOTP in KeePass using the same secret. That will let it work in all apps.

**Comment by Slummi** [comment]
KeePass has supported TOTP natively (without plugins) for some time now.
Why does KeePassXC not support the custom fields used by KeePass to store the TOTP secrets/settings?

KeePass uses the following custom fields for TOPT:

- TimeOtp-Secret (UTF-8)
- TimeOtp-Secret-Hex
- TimeOtp-Secret-Base32 (most common)
- TimeOtp-Secret-Base64

- TimeOtp-Length (optional)
- TimeOtp-Period (optional)
- TimeOtp-Algorithm (optional)

In most cases, I only have the TimeOtp-Secret-Base32 field with the secret. This works with almost all applications and plugins, but not with KeePassXC. It would be nice if KeePassXC also supported the default KeePass custom fields for TOTP.

**Reply by droidmonkey** [reply]
https://github.com/keepassxreboot/keepassxc/pull/11229

You can equally post the same question over on KeePass forums. Why did they invent new custom fields and not just use the _already existing and very efficient_ otp field?

**Reply by Slummi** [reply]
Yes, that is correct, if the native support of TOTP in KeePass was implemented after that of KeePassXC. I don't know who was faster here.
On the other hand, various other apps support the custom fields of KeePass, although they already supported TOTP before KeePass.

**Reply by droidmonkey** [reply]
We implemented it WAY before KeePass did (several years ahead). All other apps I am aware of support the otp field like we did from the extensions.

**Reply by regisdenice** [reply]
Whoever did it first, the issues stands. The TOTP secret is stored in two different fields:

- In KeePassXC: otp
- In KeePass: TimeOtp-Secret-Base32

How could we make them compatible ?
Or maybe, how could we automate a copy from one field to the other ?

**Reply by droidmonkey** [reply]
We made it compatible in 2.7.10

**Reply by regisdenice** [reply]
Great ! Thanks @droidmonkey

**Comment by regisdenice** [comment]
### Ubuntu 24.04 users

**KeePassXC v2.7.10** is not yet available (nor snap, gnome-software, apt) and I did not find it (with the keepass-cli command) in the [download page](https://github.com/keepassxreboot/keepassxc/releases/tag/2.7.10)

I'm proud to share how I installed it using the [INSTALL.md](https://github.com/keepassxreboot/keepassxc/blob/develop/INSTALL.md) documentation:

1. `sudo apt-get update && sudo apt-get upgrade`
2. `sudo apt install make build-essential cmake g++ asciidoctor qtbase5-dev qtbase5-private-dev qttools5-dev qttools5-dev-tools libqt5svg5-dev libargon2-dev libminizip-dev libbotan-2-dev libqrencode-dev libkeyutils-dev zlib1g-dev libreadline-dev libpcsclite-dev libusb-1.0-0-dev libxi-dev libxtst-dev  libqt5x11extras5-dev`
3. Download & extract "https://github.com/keepassxreboot/keepassxc/archive/refs/tags/2.7.10.tar.gz"
4. `cd 2.7.10 && mkdir build && cd build && cmake -DWITH_XC_ALL=ON .. && make`
5. `sudo make install`

Cheers

**Reply by droidmonkey** [reply]
Or you could use appimage, or flatpak, or ppa.

See https://keepassxc.org/download
